### PR TITLE
[vividus] Add new url expressions

### PR DESCRIPTION
**Pull request by PSharuba**

### CODE REVIEW

File: vividus/src/main/java/org/vividus/expression/UrlExpressionProcessors.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+@Named
+public class UrlExpressionProcessors extends DelegatingExpressionProcessor {
+
+    private static final String PATH_AND_QUERY = "%s?%s";
+
+    public UrlExpressionProcessors() {
+        super(List.of(
+                new SingleArgExpressionProcessor<>("urlHost", UrlExpressionProcessors::getUrlHost),
+                new SingleArgExpressionProcessor<>("urlPath", UrlExpressionProcessors::getUrlPath),
+                new SingleArgExpressionProcessor<>("relativeUrl", UrlExpressionProcessors::getRelativeUrl),
```
Comment (PSharuba) [inline vividus/src/main/java/org/vividus/expression/UrlExpressionProcessors.java:21]: Do we need raltiveUrl expression? In already existing steps relativeUrl check actually checks path equality. (In PageSteps.java "Then the page has the relative URL '$relativeURL'" as example)

File: docs/modules/commons/pages/expressions.adoc [hunk 2]
```diff
@@ -1026,6 +1026,47 @@ Finds the resource by its name and replaces the expression with the content of t
 Given I initialize scenario variable `my-data` with value `#{resourceToBase64(/data/body.txt)}`
 ----
 
+== Url manipulations
```
Comment (uarlouski) [inline docs/modules/commons/pages/expressions.adoc:1029]: Documentation should be written in a way not coupled to web app steps

**Comment by uarlouski** [comment]
No integration tests

File: docs/modules/commons/pages/expressions.adoc [hunk 3]
```diff
@@ -1026,6 +1026,53 @@ Finds the resource by its name and replaces the expression with the content of t
...
```
Comment (valfirst) [inline docs/modules/commons/pages/expressions.adoc:1029]: `Url manipulations` -> `URL parsing`

File: docs/modules/commons/pages/expressions.adoc [hunk 4]
```diff
@@ -1026,6 +1026,53 @@ Finds the resource by its name and replaces the expression with the content of t
...
+
+=== extractUrlHost
```
Comment (valfirst) [inline docs/modules/commons/pages/expressions.adoc:1031]:
```suggestion
=== `extractHostFromUrl`
```
please check syntax of other similar docs carefully

File: docs/modules/commons/pages/expressions.adoc [hunk 5]
```diff
@@ -1026,6 +1026,53 @@ Finds the resource by its name and replaces the expression with the content of t
...
+
+Extracts host value from given URL
```
Comment (valfirst) [inline docs/modules/commons/pages/expressions.adoc:1033]:
```suggestion
Extracts host component from the given URL.
```

File: docs/modules/commons/pages/expressions.adoc [hunk 6]
```diff
@@ -1026,6 +1026,53 @@ Finds the resource by its name and replaces the expression with the content of t
...
+
+[source, subs="+quotes"]
+----
+#{extractUrlHost(*$url*)}
```
Comment (valfirst) [inline docs/modules/commons/pages/expressions.adoc:1037]: missing documentation for the input parameters

File: docs/modules/commons/pages/expressions.adoc [hunk 7]
```diff
@@ -1026,6 +1026,53 @@ Finds the resource by its name and replaces the expression with the content of t
...
+----
+
+.Validate url host
+[source,gherkin]
+----
+Then `#{extractUrlHost(scheme://host/some/path?query)}` is = `host`
```
Comment (valfirst) [inline docs/modules/commons/pages/expressions.adoc:1043]: put some less generic examples of URL

File: docs/modules/commons/pages/expressions.adoc [hunk 8]
```diff
@@ -1026,6 +1026,53 @@ Finds the resource by its name and replaces the expression with the content of t
...
+=== extractUrlHost
+
+Extracts host value from given URL
+
+[source, subs="+quotes"]
+----
+#{extractUrlHost(*$url*)}
+----
+
+.Validate url host
```
Comment (valfirst) [inline docs/modules/commons/pages/expressions.adoc:1040]:
```suggestion
.Validate host component from URL
```

File: docs/modules/commons/pages/expressions.adoc [hunk 9]
```diff
@@ -1026,6 +1026,53 @@ Finds the resource by its name and replaces the expression with the content of t
...
+[source,gherkin]
+----
+Then `#{extractUrlHost(scheme://host/some/path?query)}` is = `host`
+----
+
+=== extractUrlPath
```
Comment (valfirst) [inline docs/modules/commons/pages/expressions.adoc:1046]: see above comments and refactor the docs

File: docs/modules/commons/pages/expressions.adoc [hunk 10]
```diff
@@ -1026,6 +1026,53 @@ Finds the resource by its name and replaces the expression with the content of t
...
+
+Extracts path value from given URL
```
Comment (valfirst) [inline docs/modules/commons/pages/expressions.adoc:1048]: is it raw path or decoded path?

File: docs/modules/commons/pages/expressions.adoc [hunk 11]
```diff
@@ -1026,6 +1026,53 @@ Finds the resource by its name and replaces the expression with the content of t
...
+
+.Validate url path
+[source,gherkin]
+----
+Then `#{extractUrlPath(scheme://host/some/path?query)}` is = `/some/path`
+----
+
+=== extractUrlQuery
+
+Extracts query value from given URL
```
Comment (valfirst) [inline docs/modules/commons/pages/expressions.adoc:1063]: is it raw query or decoded query?

File: vividus-tests/src/main/resources/story/integration/Expressions.story [hunk 12]
```diff
@@ -217,3 +217,11 @@ Examples:
 
 Scenario: Validate Data URL decoding expression for binary data
 When I compare against baseline with name `black-pixel` from image `#{decodeDataUrl(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAAAXNSR0IArs4c6QAAAA1JREFUGFdjYGBg+A8AAQQBAHAgZQsAAAAASUVORK5CYII=)}`
+
+Scenario: Validate url expressions
```
Comment (valfirst) [inline vividus-tests/src/main/resources/story/integration/Expressions.story:221]:
```suggestion
Scenario: Validate expressions parsing URLs
```

File: vividus-tests/src/main/resources/story/integration/Expressions.story [hunk 13]
```diff
@@ -217,3 +217,11 @@ Examples:
...
+Then `#{<expression>}` is = `<expected>`
+Examples:
+|expression                                                       |expected                  |
+|extractUrlHost(${vividus-test-site-url}/windows.html?query=test) |${vividus-test-site-host} |
+|extractUrlPath(${vividus-test-site-url}/windows.html?query=test) |/windows.html             |
+|extractUrlQuery(${vividus-test-site-url}/windows.html?query=test)|query=test                |
```
Comment (valfirst) [inline vividus-tests/src/main/resources/story/integration/Expressions.story:227]: add examples with encoded path and encoded query

File: vividus/src/main/java/org/vividus/expression/UrlExpressionProcessors.java [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+import java.util.List;
+
+import javax.inject.Named;
+
+import org.apache.commons.lang3.Validate;
+import org.jbehave.core.expressions.DelegatingExpressionProcessor;
+import org.jbehave.core.expressions.SingleArgExpressionProcessor;
+import org.vividus.util.UriUtils;
+
+@Named
```
Comment (valfirst) [inline vividus/src/main/java/org/vividus/expression/UrlExpressionProcessors.java:28]: register bean explicitly

File: vividus/src/main/java/org/vividus/expression/UrlExpressionProcessors.java [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+    {
+        String urlHost = UriUtils.createUri(url).getHost();
+        Validate.notNull(urlHost, "URL host component of '%s' is null", url);
+        return urlHost;
+    }
+
+    private static String extractUrlPath(String url)
+    {
+        String urlPath = UriUtils.createUri(url).getPath();
+        Validate.notNull(urlPath, "URL path component of '%s' is null", url);
```
Comment (valfirst) [inline vividus/src/main/java/org/vividus/expression/UrlExpressionProcessors.java:48]: why is this validation needed?

File: vividus/src/main/java/org/vividus/expression/UrlExpressionProcessors.java [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+        return urlPath;
+    }
+
+    private static String extractUrlQuery(String url)
+    {
+        String urlQuery = UriUtils.createUri(url).getQuery();
+        Validate.notNull(urlQuery, "URL query component of '%s' is null", url);
```
Comment (valfirst) [inline vividus/src/main/java/org/vividus/expression/UrlExpressionProcessors.java:55]: why is this validation needed?

File: vividus/src/test/java/org/vividus/expression/UrlExpressionProcessorsTests.java [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+@ExtendWith(MockitoExtension.class)
+class UrlExpressionProcessorsTests
+{
+    private static final String EXAMPLE_URL = "https://www.example.com/one/two?id=0&name=temp";
+    private static final String URL_HOST = "www.example.com";
+    private static final String URL_PATH = "/one/two";
+    private static final String URL_QUERY = "id=0&name=temp";
+    private static final String VALIDATION_ERROR_MESSAGE = "URL %s component of '%s' is null";
+
+    @InjectMocks private UrlExpressionProcessors processors;
```
Comment (valfirst) [inline vividus/src/test/java/org/vividus/expression/UrlExpressionProcessorsTests.java:41]: nothing to inject

**Review by valfirst — CHANGES_REQUESTED** [review]

File: docs/modules/commons/pages/expressions.adoc [hunk 18]
```diff
@@ -1026,6 +1026,66 @@ Finds the resource by its name and replaces the expression with the content of t
 Given I initialize scenario variable `my-data` with value `#{resourceToBase64(/data/body.txt)}`
 ----
 
+== URL parsing
+:url-parameter: pass:quotes[*`$url`* - the URL to extract from]
```
Comment (valfirst) [inline docs/modules/commons/pages/expressions.adoc:1030]:
```suggestion
:url-parameter: pass:quotes[*`$url`* - the URL to extract component from.]
```

**Review by valfirst — CHANGES_REQUESTED** [review]

**Pull request merged**